File: scripts/extract_thumbs.py
#!/usr/bin/env python3
"""
视频首帧自动提取脚本（使用 imageio-ffmpeg 捆绑的 ffmpeg 二进制）
零系统依赖，自动通过 pip 安装 imageio-ffmpeg 即可获得 ffmpeg。

用法：
  python scripts/extract_thumbs.py

也可在 Node.js 中通过 child_process 调用。
"""

import os
import sys
import subprocess
from pathlib import Path

# 项目根目录（脚本位于 scripts/）
PROJECT_ROOT = Path(__file__).resolve().parent.parent
VIDEO_DIR = PROJECT_ROOT / "public" / "VIDEO"
THUMBS_DIR = VIDEO_DIR / "thumbs"


def get_ffmpeg():
    """获取 ffmpeg 可执行文件路径（优先系统，其次 imageio-ffmpeg 捆绑）"""
    # 先尝试系统 ffmpeg
    try:
        subprocess.run(["ffmpeg", "-version"], capture_output=True, timeout=5)
        return "ffmpeg"
    except (FileNotFoundError, subprocess.TimeoutExpired):
        pass

    # 尝试 imageio-ffmpeg 捆绑的 ffmpeg
    try:
        import imageio_ffmpeg
        ffmpeg_exe = imageio_ffmpeg.get_ffmpeg_exe()
        print(f"[Thumbs/Python] 使用 imageio-ffmpeg 捆绑的 ffmpeg: {ffmpeg_exe}")
        return ffmpeg_exe
    except ImportError:
        pass

    return None


def extract_frame(video_path, thumb_path, ffmpeg):
    """使用 ffmpeg 提取视频首帧"""
    try:
        result = subprocess.run(
            [
                ffmpeg,
                "-y",           # 覆盖已存在文件
                "-ss", "00:00.5",  # 跳转到 0.5 秒
                "-i", str(video_path),
                "-vframes", "1",   # 只提取 1 帧
                "-q:v", "2",       # 高质量 JPEG
                "-loglevel", "error",
                str(thumb_path),
            ],
            capture_output=True,
            timeout=60,
        )
        return result.returncode == 0 and thumb_path.exists()
    except Exception as e:
        print(f"  [Thumbs/Python] 提取失败: {e}")
        return False


def main():
    print("[Thumbs/Python] ===== 视频缩略图提取 =====")

    # 获取 ffmpeg
    ffmpeg = get_ffmpeg()
    if not ffmpeg:
        print("[Thumbs/Python] ❌ 未找到 ffmpeg！")
        print("[Thumbs/Python]    请安装: pip install imageio-ffmpeg")
        sys.exit(1)

    # 确保目录存在
    if not VIDEO_DIR.exists():
        print(f"[Thumbs/Python] ❌ 视频目录不存在: {VIDEO_DIR}")
        sys.exit(1)

    THUMBS_DIR.mkdir(parents=True, exist_ok=True)

    # 扫描视频文件
    video_files = sorted(
        [f for f in VIDEO_DIR.glob("*.mp4") if f.stem.isdigit()],
        key=lambda f: int(f.stem)
    )

    if not video_files:
        print("[Thumbs/Python] ⚠ 未找到编号命名的视频文件（N.mp4）")
        return {"total": 0, "extracted": 0, "skipped": 0, "failed": 0}

    print(f"[Thumbs/Python] 找到 {len(video_files)} 个视频文件")

    extracted = 0
    skipped = 0
    failed = 0

    for vf in video_files:
        num = int(vf.stem)
        thumb_path = THUMBS_DIR / f"{num}.jpg"

        if thumb_path.exists():
            skipped += 1
            continue

        print(f"[Thumbs/Python] 提取 {vf.name} 首帧...", end=" ", flush=True)

        if extract_frame(vf, thumb_path, ffmpeg):
            size_kb = thumb_path.stat().st_size / 1024
            print(f"✅ {size_kb:.1f} KB")
            extracted += 1
        else:
            print("❌ 失败")
            failed += 1

    print(f"[Thumbs/Python] ===== 完成: 跳过 {skipped} | 提取 {extracted} | 失败 {failed} =====")
    return {"total": len(video_files), "extracted": extracted, "skipped": skipped, "failed": failed}


if __name__ == "__main__":
    sys.exit(0 if main()["failed"] == 0 else 1)

File: scripts/extract-thumbs.js
/**
 * 视频首帧手动提取工具（备选方案，需要 ffmpeg）
 *
 * 主方案：extract-thumbs-plugin.js（Vite 插件，零依赖，浏览器 Canvas API）
 * 备选方案：此脚本（命令行，需要安装 ffmpeg）
 *
 * 用法：node scripts/extract-thumbs.js
 *
 * 前置条件：系统已安装 ffmpeg 或在 node_modules 中安装了 ffmpeg-static
 */

const { execSync, spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

const VIDEO_DIR = path.resolve(__dirname, '..', 'public', 'VIDEO');
const THUMBS_DIR = path.join(VIDEO_DIR, 'thumbs');

function log(msg) {
  console.log(`[Thumbs/manual] ${msg}`);
}

function getFfmpegPath() {
  try { return require('ffmpeg-static'); } catch {}
  try { execSync('ffmpeg -version', { stdio: 'pipe' }); return 'ffmpeg'; } catch {}
  return null;
}

function extractThumb(videoPath, thumbPath, ffmpeg) {
  return new Promise((resolve) => {
    const proc = spawn(ffmpeg, [
      '-y', '-ss', '00:00.5', '-i', videoPath,
      '-vframes', '1', '-q:v', '2', '-loglevel', 'error',
      thumbPath,
    ]);
    let stderr = '';
    proc.stderr.on('data', d => { stderr += d.toString(); });
    proc.on('close', code => resolve(code === 0 && fs.existsSync(thumbPath)));
    proc.on('error', () => resolve(false));
  });
}

async function main() {
  log('===== 视频缩略图手动提取（ffmpeg）=====');

  const ffmpeg = getFfmpegPath();
  if (!ffmpeg) {
    log('❌ 未找到 ffmpeg。请使用 Vite 插件的浏览器自动提取方案。');
    log('   启动 dev server: npm run dev');
    process.exit(1);
  }

  log(`ffmpeg: ${ffmpeg}`);

  if (!fs.existsSync(THUMBS_DIR)) fs.mkdirSync(THUMBS_DIR, { recursive: true });

  const files = fs.readdirSync(VIDEO_DIR)
    .filter(f => /^\d+\.mp4$/i.test(f))
    .map(f => parseInt(f.match(/^(\d+)/)[1]))
    .sort((a, b) => a - b);

  let extracted = 0, skipped = 0;
  for (const num of files) {
    const thumbPath = path.join(THUMBS_DIR, `${num}.jpg`);
    if (fs.existsSync(thumbPath)) { skipped++; continue; }

    const videoPath = path.join(VIDEO_DIR, `${num}.mp4`);
    log(`提取 ${num}.mp4...`);
    const ok = await extractThumb(videoPath, thumbPath, ffmpeg);
    if (ok) { extracted++; log(`  ✅ ${num}.jpg`); }
    else { log(`  ❌ 失败`); }
  }

  log(`===== 完成: 跳过 ${skipped} | 提取 ${extracted} =====`);
  process.exit(extracted === files.length - skipped ? 0 : 1);
}

main().catch(e => { log(`错误: ${e.message}`); process.exit(1); });
